### PR TITLE
[zk-sdk] Expose `ElGamalPubkey` and `ElGamalKeypair` as wasm

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
There are no js handles to do cryptographic operations in `zk-sdk`. Instead of implementing cryptographic operations in js, we can compile the logic to target wasm.

#### Summary of Changes
Added dependency to `wasm-bindgen` and exposed `ElGamalPubkey` and `ElGamalKeypair` in wasm target. For these types, I generally followed how it is done in the solana-sdk with `Pubkey` and `Keypair`. For the `ElGamalKeypair` type, I also added a function `fn pubkey_owned(&self) -> ElGamalPubkey` since the standard `fn pubkey(&self) -> &ElGamalPubkey` does not work in wasm since it is returning a reference.

[002808a](https://github.com/anza-xyz/agave/pull/2996/commits/002808a0d8b55e653fd552b31da89879eb36ae22): I added `wasm-bindgen` dependency. I explicitly excluded `solana-sdk` from wasm target since it blows up the wasm binary with types like `Pubkey`, `Keypair`, and `Instructions` that are not really needed for `zk-sdk`. I also excluded the `zk-elgamal-program` module for now since the `solana-program` dependency also seems to blow up the wasm binary.

[8923382](https://github.com/anza-xyz/agave/pull/2996/commits/89233825dcb904cdc7c03449f684b1a94b3b1d22): I exposed `ElGamalPubkey` and `ElGamalKeyapir` as `wasm_bindgen`. I tried exposing the `PodElGamalPubkey` type also using `wasm_bindgen`, but it seemed that the solana target and wasm target did not play well and I ended up defining a new type `CompressedElGamalPubkey` for the compactly encoded `ElGamalPubkey` type.

This PR is a replacement and a much simpler version of https://github.com/anza-xyz/agave/pull/333.

One can build using the command `wasm-pack build --target nodejs`. The wasm binary seems to be about 95kb.

Fixes #

### CODE REVIEW

File: zk-sdk/Cargo.toml
Comment (joncinque) [inline zk-sdk/Cargo.toml]: Is this a separate change?

File: zk-sdk/Cargo.toml
Comment (joncinque) [inline zk-sdk/Cargo.toml]: What do you think about only requiring the wasm dependencies if the target is wasm, similar to what we're doing at https://github.com/anza-xyz/agave/blob/83e7d84bcc4cf438905d07279bc07e012a49afd9/sdk/pubkey/Cargo.toml#L40-L43

File: zk-sdk/src/encryption/auth_encryption.rs
Comment (joncinque) [inline zk-sdk/src/encryption/auth_encryption.rs]: What's the reason to exclude some of these functions? I can imagine that `&dyn Signer` doesn't make sense, but this one seems reasonable

File: zk-sdk/src/encryption/auth_encryption.rs
Comment (joncinque) [inline zk-sdk/src/encryption/auth_encryption.rs:162]: So just so I understand, these are *not* included because it makes the wasm binary too big, correct? if so, can you add a comment to that effect? otherwise we might accidentally add it in the future

File: zk-sdk/src/encryption/elgamal.rs
Comment (joncinque) [inline zk-sdk/src/encryption/elgamal.rs]: Can we only pull this in for wasm builds, similar to https://github.com/anza-xyz/agave/blob/83e7d84bcc4cf438905d07279bc07e012a49afd9/sdk/pubkey/src/lib.rs#L31-L35

File: zk-sdk/src/encryption/elgamal.rs
Comment (joncinque) [inline zk-sdk/src/encryption/elgamal.rs:175]: nit: Would it make more sense to have this whole impl block under `#[cfg(not(target_arch = "wasm32"))]`?

File: zk-sdk/src/encryption/elgamal.rs
Comment (joncinque) [inline zk-sdk/src/encryption/elgamal.rs]: nit: I can't comment further up, but similar to the last one, would it make more sense to have this whole impl block take the `#[cfg(not(target_arch = "wasm32"))]` attribute? And then you can put all of the functions that are ok for wasm in a separate impl block, which might look cleaner

File: zk-sdk/src/encryption/wasm/elgamal.rs
Comment (joncinque) [inline zk-sdk/src/encryption/wasm/elgamal.rs:1]: Is it possible to avoid this by adding `#[cfg(target_arch = "wasm32")]` in the right places and adding specific impl blocks for wasm32? There's a good example of that at https://github.com/anza-xyz/agave/blob/83e7d84bcc4cf438905d07279bc07e012a49afd9/sdk/pubkey/src/lib.rs#L907-L910

**Review by joncinque — COMMENTED** [review]
Looks good overall, let's see if we can avoid the copy-pasta new type and just keep it in the same type. And let's protect with `target_arch = "wasm32"` as much as possible, so that we don't have use wasm bindgen all the time

File: zk-sdk/Cargo.toml
Comment (samkim-crypto) [reply zk-sdk/Cargo.toml]: Yeah this was put in initially as specified in the bindgen docs to minimize the wasm binary size as much as possible. It would be nice to be able to add a target specific profile in cargo, but it seems that it is not possible. 

For the changes in this PR though, adding this profile seems to have no impact on the binary size, so I removed it. For subsequent PRs, I will experiment with adding this profile to see if it affects the binary size.

File: zk-sdk/Cargo.toml
Comment (samkim-crypto) [reply zk-sdk/Cargo.toml]: Ah yes, I should have done this initially. Thanks!

File: zk-sdk/src/encryption/auth_encryption.rs
Comment (samkim-crypto) [reply zk-sdk/src/encryption/auth_encryption.rs]: The problem is that if I include `solana-sdk` as part of the wasm target build, then the resulting wasm binary seems to include types like `Pubkey` or `Signature` that are not specific to the `zk-sdk` (https://github.com/rustwasm/wasm-bindgen/issues/3759). So I manually added `#[cfg(not(target_arch = "wasm32"))]` to all the places that depend on `solana-sdk` to prevent unnecessary bloat.

It might make sense to export this function for wasm targets later on (and in that case, it seems that we would need to depend on `solana-sdk`), but at least for this PR, I excluded it.

File: zk-sdk/src/encryption/auth_encryption.rs
Comment (samkim-crypto) [reply zk-sdk/src/encryption/auth_encryption.rs:162]: That's right. It is to prevent bloat in the binary. I added a comment where we import the dependencies. If we do accidentally remove this attribute in the future, it will not compile since `solana-sdk` is set to `#[cfg(not(target_arch = "wasm32"))]`.

File: zk-sdk/src/encryption/elgamal.rs
Comment (samkim-crypto) [reply zk-sdk/src/encryption/elgamal.rs:175]: Yep good idea. Done!

File: zk-sdk/src/encryption/grouped_elgamal.rs
Comment (joncinque) [inline zk-sdk/src/encryption/grouped_elgamal.rs]: nit: can you add these in a separate impl block as in the other file?

File: zk-sdk/src/encryption/grouped_elgamal.rs
Comment (joncinque) [inline zk-sdk/src/encryption/grouped_elgamal.rs]: Same with these, can they go in a separate impl block?

File: zk-sdk/src/encryption/wasm/elgamal.rs
Comment (joncinque) [inline zk-sdk/src/encryption/wasm/elgamal.rs]: Rather than adding the new type, can we just add a separate impl block for `PodElGamalPubkey`? I got this working on my side, do you mind if I push to your branch so you can see what it looks like?

Then we'll also be able to eliminate all of the `wasm` modules!

**Review by joncinque — COMMENTED** [review]
Sorry for the delay here, just a few more comments, then this is good to go!

File: zk-sdk/src/encryption/grouped_elgamal.rs
Comment (samkim-crypto) [reply zk-sdk/src/encryption/grouped_elgamal.rs]: Sorry I missed these!

File: zk-sdk/src/encryption/wasm/elgamal.rs
Comment (samkim-crypto) [reply zk-sdk/src/encryption/wasm/elgamal.rs]: Oh yes, I couldn't quite get the `PodElGamalPubkey` to work in both solana and wasm targets, but if you got it work, then please feel free to push the changes!

Comment (joncinque) [reply zk-sdk/src/encryption/wasm/elgamal.rs]: Ok great, let me know how [0830405](https://github.com/anza-xyz/agave/pull/2996/commits/0830405251c16953dd3b20901ab1692bb380731b) looks!

**Comment by joncinque** [comment]
All good on my side, so once you approve my changes, we can merge this in!

**Review by samkim-crypto — COMMENTED** [review]
Hm, I think I tried these changes before, but for some reason I couldn't quite satisfy the CI then, maybe due to the use of `display_to_jsvalue`. But it is good that it is all figured out. Thanks a lot!

I remember encountering this problem before (and figured it back then), but github does not allow PR authors to do any approvals even for contributions by another author. Do you have a pointer on how to proceed with the approval 😅 ?

**Review by joncinque — APPROVED** [review]

**Comment by joncinque** [comment]
> Hm, I think I tried these changes before, but for some reason I couldn't quite satisfy the CI then, maybe due to the use of display_to_jsvalue. But it is good that it is all figured out. Thanks a lot!

Ok great!

> I remember encountering this problem before (and figured it back then), but github does not allow PR authors to do any approvals even for contributions by another author. Do you have a pointer on how to proceed with the approval 😅 ?

Uh oh, I can't remember the workaround. Maybe you can rebase and push, and that'll allow me to approve?

**Comment by samkim-crypto** [comment]
I just rebased, but it seems like your approval remains and I can't seem to do any approvals. Looking back, I think we encountered this problem for a backport PR where the git bot was the PR author. We might need someone else to approve this PR for us...?

**Review by yihau — APPROVED** [review]
🪖

**Pull request merged**
